drivers/i2c/i2c_ll_stm32.c

```diff
@@ -73,15 +73,23 @@ int i2c_stm32_runtime_configure(const struct device *dev, uint32_t config)
 	k_sem_take(&data->bus_mutex, K_FOREVER);
 
 #ifdef CONFIG_PM_DEVICE_RUNTIME
-	(void)pm_device_runtime_get(dev);
```

Comment:
I generally agree with your sentiment, pinctrl isn't necessary to configure the bus.
However, the `pm_device_runtime_get/put` functions also operate as a sort of semaphore with the usage field, so if the usage counter is already >1 when we reach this section, we'll unnecessarily be toggling the clock where the runtime get/put wouldn't.

Reply:
> so if the usage counter is already >1 when we reach this section

Meaning that there would already be either a `i2c_stm32_runtime_configure()` or `i2c_stm32_transfer()` on going when this call to `i2c_stm32_runtime_configure()` happens. This is possible but unlikely and I think that the kind of case that should be avoided from applicative side anyway.

OTOH, it is likely that `i2c_stm32_runtime_configure()` happens while counter is 0, triggering the whole machinery of pinctrl configuration/deconfiguration, which it self generates 2 GPIO clock_on/clock_off best case (worst case, `n` GPIO controllers are involved , so `n` time 2 GPIO clock_on/clock_off).

Reply:
> This is possible but unlikely and I think that the kind of case that should be avoided from applicative side anyway.

It still disrupts/bypasses the runtime get/put pattern in a way that may lead to unintended side effects. This doesn't rub me the right way from an architectural point of view.

> while counter is 0, triggering the whole machinery of pinctrl configuration/deconfiguration, which it self generates 2 GPIO clock_on/clock_off

I agree with you, I'm just wondering if there's a more streamlined approach to this with runtime PM. I doubt this is the only driver where runtime PM actions cascade into other peripheral actions unnecessarily.

Reply:
> It still disrupts/bypasses the runtime get/put pattern in a way that may lead to unintended side effects. This doesn't rub me the right way from an architectural point of view.

One case might happen that we're in the middle of `i2c_stm32_runtime_configure()` and `clock_control_on()` was already called while a higher prio thread is calling `i2c_stm32_transfer()`. 
Fortunately we're protected of this case by the `bus_mutex` semaphore, which will ensure `i2c_stm32_runtime_configure()` is completed and `clock_control_off()` called before `i2c_stm32_transfer()` and `pm_device_runtime_get()` happens. So I don't see things goign really bad in any case.

Reply:
You're right, the `bus_mutex` will always guard this and transactions sections, regardless. From a design point of view it feels like we're breaking the device runtime contract directly toggling peripheral clocks, but there's no concrete reason current why we shouldn't do this.